tester/include/game_tester/strategy.hpp

```diff
@@ -49,37 +50,40 @@ namespace testing::strategy {
     };
 
     struct Graph {
-        std::shared_ptr<Node> _root;
+        explicit Graph(action_t && root_action) : root(std::make_shared<strategy::Node>(std::move(root_action))) {
+        }
+
+        std::shared_ptr<Node> root;
     };
 
     class Executor {
     public:
         explicit Executor(Graph && graph) : _graph(graph) {
-            _current = graph._root;
+            _current = graph.root;
         }
 
         uint process_strategy(game_tester & tester,
                               const uint run_count,
                               const uint limit_per_run,
-                              std::function<void(game_tester &, const uint)> && pre_run_callback,
-                              std::function<void(const game_tester &)> && post_run_callback) {
+                              std::function<session_id_t(game_tester &, const uint)> && session_create,
+                              std::function<void(game_tester &, const session_id_t session_id)> && session_close) {
 
             for (uint run = 0; run != run_count; ++run) {
-                pre_run_callback(tester, run);
+                const auto session_id = session_create(tester, run);
 
-                if (!process_run(tester, limit_per_run))
+                if (!process_run(tester, session_id, limit_per_run))
                     return run;
 
-                post_run_callback(tester);
+                session_close(tester, session_id);
             }
 
             return run_count;
         }
 
     private:
-        bool process_run(game_tester & tester, uint limit) {
+        bool process_run(game_tester & tester, const session_id_t session_id, uint limit) {
```

Comment:
I think it's better to init `_current` here. This way we would worry less and wouldn't forget to reset it deeply in `process_next_step`.

Reply:
But `current` always initialized. If `current == nullptr => it was error.